### ISSUE TITLE
Usability issues with http / https register

### ISSUE THREAD

**Issue by jerstlouis** [opengeospatial/NamingAuthority#316]
This **https** URL:

https://www.opengis.net/def/interpolation/OGC/0/

(from https://github.com/opengeospatial/NamingAuthority/issues/260 )

is broken in at least 2 ways:

- it redirects to an **http** URI
- it results in an Input Error

```
wget https://www.opengis.net/def/interpolation/OGC/0/
--2024-12-11 11:51:29--  https://www.opengis.net/def/interpolation/OGC/0/
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving www.opengis.net (www.opengis.net)... 66.244.86.70
Connecting to www.opengis.net (www.opengis.net)|66.244.86.70|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: http://defs.opengis.net/vocprez/object?uri=https://www.opengis.net/def/interpolation/OGC/0/&_mediatype=text/turtle [following]
--2024-12-11 11:51:30--  http://defs.opengis.net/vocprez/object?uri=https://www.opengis.net/def/interpolation/OGC/0/&_mediatype=text/turtle
Resolving defs.opengis.net (defs.opengis.net)... 66.244.86.71
Connecting to defs.opengis.net (defs.opengis.net)|66.244.86.71|:80... connected.
HTTP request sent, awaiting response... 400 BAD REQUEST
2024-12-11 11:51:30 ERROR 400: BAD REQUEST.
```

Shouldn't `/0` always resolve to the latest?

https://www.opengis.net/def/interpolation/OGC/1/

is also broken on Chrome.

It redirects to: 

http://defs.opengis.net/vocprez/object?uri=https://www.opengis.net/def/interpolation/OGC/1/

when it should probably redirect instead to:

https://defs.opengis.net/vocprez/object?uri=http://www.opengis.net/def/interpolation/OGC/1/

I still believe that the concept of "alias register" for http / https in the definition server is causing all kinds of usability troubles (e.g. landing on a completely different web page for each, and even different page using different browsers for the same URI), and simply treating https / http URIs as equivalent for the purpose of lookups.

The redirects to `/vocprez` changing URIs entered in the URIs still contribute to painful user experience with the definition server (#85 #109), especially when combined when trying to understand such problems.

**Comment by rob-metalinkage** [comment]
"when it should probably redirect instead to:

https://defs.opengis.net/vocprez/object?uri=http://www.opengis.net/def/interpolation/OGC/1/"

I believe this is correct - @kstegemoller can you fix the top level redirects - I think this is done there from memory and from what I can tell looking at the defs server config.  @avillar please confirm. 

IMHO we probably need a Naming Authority action to define functionality tests.  I am planning to revisit the redirection and content negotiation infrastructure next year when I can sit down with @avillar and explore simplifications possible now we have more extensive github automations available.

**Comment by avillar** [comment]
There's two things at play here:

1) I'm pretty sure the https -> http redirection is done by Django (uriredirect?). I can definitely see the request to https://www.opengis.net/def/interpolation/OGC/0/ being handled by the Django instance in the logs. So unless I'm mistaken it's a matter of configuring the URI redirects to use the https VocPrez for visualization.

2) Nonexistent resources will result in a 400 or 500 error from VocPrez (this is something that we'll fix in the upcoming RAINBOW version). **https**://www.opengis.net/def/interpolation/OGC/0/ doesn't exist (checked against triplestore), and neither do **http**://www.opengis.net/def/interpolation/OGC/0/ or **https**://www.opengis.net/def/interpolation/OGC/1/. The only one that contains any data is **http**://www.opengis.net/def/interpolation/OGC/1 (which has no https alias!).

**Comment by ghobona** [comment]
There is no resource registered at https://www.opengis.net/def/interpolation/OGC/0/ . We could not find a standard that uses that URI. Please let us know if you know of one.

There is, however, a resource registered at http://www.opengis.net/def/interpolation/OGC/1/ . This URI is originally from the WCS Interpolation Extension standard [(OGC 12-049)](https://portal.ogc.org/files/12-049).

**Comment by jerstlouis** [comment]
Thanks @ghobona 

Shouldn't `/0/` always refer to the latest available version, or be used for identifiers where a version is not applicable? New concepts being defined should probably just be added to `/0/` as opposed to changing all URIs everytime we want to add a new concept. ( https://docs.ogc.org/pol/09-048r6.html#toc13 https://github.com/opengeospatial/NamingAuthority/issues/143 )

Whereas it might sense to have versioned standards, most conceptual registers should not be versioned I think.

e.g., QUDT definitions like https://qudt.org/vocab/quantitykind/AtmosphericPressure have no "versions".

**Comment by pebau** [comment]
> There is, however, a resource registered at http://www.opengis.net/def/interpolation/OGC/1/ . This URI is originally from the WCS Interpolation Extension standard [(OGC 12-049)](https://portal.ogc.org/files/12-049).

FWIW, the WCS Interpolation extension does not have any particular requirements on such URIs, in particular it does not normatively require that path. It was just an example based on the recommendation by OGC people at that time. 

Bottom line, from a WCS perspective it can be tweaked if necessary.

**Comment by jerstlouis** [comment]
@chris-little I am actually of a somewhat opposite opinion... I believe many concepts like link relations, 2DTMS / DGGRS definitions, should "never change" and they should not have a version number at all. The only reason for them to change is if they were broken in the first place and did not work, or if their representation change, which should not require an identifier change but representation negotiation mechanism instead. That is somewhat different from `/0/` as the latest version rather than _Version Not Applicable_, but related.

**Comment by dr-shorthair** [comment]
>  for major releases, that are not backward compatible

Major changes to definitions should have new URIs. Always.

**Comment by avillar** [comment]
I've come up with a new solution to the http/https issue that doesn't depend on the alias register. Whenever an **https** resource URI is resolved by the backend we check whether:

- it doesn't exist (which would result in a 404), but an http version does.
- it has an `owl:sameAs` to an `http://` version of itself.

If any of those is true:

- for HTML views, we redirect the uset to the UI description of the **http** version.
- for RDF data views, we redirect to the data endpoint of the **http** version.

I've checked our data, and all https resources that have an http sameAs belong to the alias register, so there shouldn't be any undesired side effects.

I've enabled this on defs-dev for now, so no direct URI resolution, but we can test it like so:

## HTML view
```
curl "https://defs-dev.opengis.net/prez-b/object?uri=https://www.opengis.net/def/iso-8601/Gregorian" -v 2>&1 | grep -i location:
< location: https://defs-dev.opengis.net/prez/object?uri=http%3A%2F%2Fwww.opengis.net%2Fdef%2Fiso-8601%2FGregorian
```

## Data view
```
curl "https://defs-dev.opengis.net/prez-b/object?uri=https://www.opengis.net/def/iso-8601/Gregorian" -v -H 'Accept: text/turtle' 2>&1 | grep -i location:
< location: https://defs-dev.opengis.net/prez-b/object?uri=http%3A%2F%2Fwww.opengis.net%2Fdef%2Fiso-8601%2FGregorian
```

## Data view (accepting redirecton)
(The description is for the http version)
```
curl "https://defs-dev.opengis.net/prez-b/object?uri=https://www.opengis.net/def/iso-8601/Gregorian" -H 'Accept: text/turtle' -L
@prefix iso8601: <http://www.opengis.net/def/iso-8601/> .
@prefix ns-1904448668017821819: <http://www.opengis.net/def/uom/ISO-8601/0/> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix prez: <https://prez.dev/> .
@prefix profile: <https://prez.dev/profile/> .
@prefix skos: <http://www.w3.org/2004/02/skos/core#> .

iso8601:Gregorian owl:sameAs ns-1904448668017821819:Gregorian ;
    skos:prefLabel "Alias of http://www.opengis.net/def/iso-8601/Gregorian"@en,
        "Gregorian"@en ;
    prez:label "Alias of http://www.opengis.net/def/iso-8601/Gregorian"@en,
        "Gregorian"@en ;
[...]
```

**Comment by jerstlouis** [comment]
@avillar The definition server still seems to be returning errors for:

```
curl -v https://www.opengis.net/def/interpolation/OGC/1
```

Which 303 redirects to:

 Location: https://defs.opengis.net/prez-backend/object?uri=https://www.opengis.net/def/interpolation/OGC/1

Then:

```
curl -v https://defs.opengis.net/prez-backend/object?uri=https://www.opengis.net/def/interpolation/OGC/1

< HTTP/1.1 307 Temporary Redirect
< Date: Tue, 12 May 2026 22:07:51 GMT
< Server: uvicorn
< content-length: 0
< location: https://defs.opengis.net/prez/404?uri=https%3A%2F%2Fwww.opengis.net%2Fdef%2Finterpolation%2FOGC%2F1
```

**Comment by avillar** [comment]
Precisely! The production one doesn't have my fix, so it doesn't do the behind-the-scenes https->http rewrite. I'll push the change to prod tomorrow morning (I'd rather not deal with any potential issues tonight :) ) and we can test it.

**Comment by avillar** [comment]
Enabled on prod, and it seems to be working as intended. One thing though: we're only checking `https` URLs for 404/sameAs to `http`, but if the incoming request uses `http` directly, we don't do any sort of check.

Maybe we should apply the same logic to plain `http` resources, so if a user mistakenly requests an `https` one using `http`, we'll serve the appropriate content. However, this is much less likely to happen on accident (since I expect browsers to upgrade connections automatically, but not to downgrade them), but the possibility is there.
